File: ruby/roman_to_int.rb
# @param {String} s
# @return {Integer}
# within 1-3999

def roman_to_int(s)
    idx = 0
    result = 0
    while idx < s.length
      if s.slice(idx, 2) === 'IV'
        result += 4
        idx+=1
      elsif s.slice(idx, 2) === 'IX'
        result += 9
        idx+=1
      elsif s.slice(idx, 2) === 'XL'
        result += 40
        idx+=1
      elsif s.slice(idx, 2) === 'XC'
        result += 90
        idx+=1
      elsif s.slice(idx, 2) === 'CD'
        result += 400
        idx+=1
      elsif s.slice(idx, 2) === 'CM'
        result += 900
        idx+=1
      elsif s[idx] === 'I'
        result += 1
      elsif s[idx] === 'V'
        result += 5
      elsif s[idx] === 'X'
        result += 10
      elsif s[idx] === 'L'
        result += 50
      elsif s[idx] === 'C'
        result += 100
      elsif s[idx] === 'D'
        result += 500
      elsif s[idx] === 'M'
        result += 1000   
      end
      idx+=1
    end
    result
end

one = 'III'
two = 'IV'
three = 'IX'
four = 'LVIII'
five = 'MCMXCIV'

# p 'one'
# p roman_to_int(one) # * 3
# p 'two'
# p roman_to_int(two) # * 4
# p roman_to_int(three) # * 9
# p roman_to_int(four) # * 58
p 'five'
p roman_to_int(five) # * 1994

File: fis/roman_to_int.js
/**
 * @param {string} s
 * @return {number}
 *
 * Given a string of Roman Numerals
 * the output is their value as an integer
 *
 */

const romNums = {
  I: 1,
  V: 5,
  X: 10,
  L: 50,
  C: 100,
  D: 500,
  M: 1000,
}

const specials = {
  IV: 4,
  IX: 9,
  XL: 40,
  XC: 90,
  CD: 400,
  CM: 900,
}

const romanToIntFirst = (s) => {
  // We need to split the array into characters
  // Create a starting total of 0
  // First thought is to loop over the array and apply the values for each numeral
  // Inside the loop:
  // increase total by the value for the string
  // ex: III => I(1) + I(1) + I(1) = 3
  // romArr.forEach((numeral) => {
  //   total += romNums[numeral]
  // })
  // return total

  // How to account for the IV, IX, CM...
  // first check if the current pair of elements (i and i + 1) are our special cases, if so then we increase the total by the special count, and move our index forward 1
  // MCMXCIV  => 1994
  // M(1000) + CM(900) + XC(90) + IV(4)

  // ? Another way for thes special cases
  // Inside the loop:
  // If the NEXT value is GREATER THAN our current value,
  // Then instead of finding the special pair,
  // we increase the total by NEXT - CURRENT and the counter by 1
  // EX: MCMXCIV  => 1994
  // M(1000) + (M(1000) - C(100)) + ()

  const romArr = s.split("")
  let total = 0
  let i = 0
  while (i < romArr.length) {
    let element = romArr[i]
    let pair = romArr.slice(i, i + 2).join("")
    console.log(pair)
    if (!!specials[pair]) {
      total += specials[pair]
      i++
    } else {
      total += romNums[element]
    }
    i++
  }
  return total
}

// TESTING

let s = "III"
let ans = 3
// console.log(romanToIntFirst(s), ans)
// s = "IV"
// ans = 4
// console.log(romanToIntFirst(s), ans)
// s = "IX"
// ans = 9
// console.log(romanToIntFirst(s), ans)
// s = "LVIII"
// ans = 58
// console.log(romanToIntFirst(s), ans)
// s = "MCMXCIV"
// ans = 1994
// console.log(romanToIntFirst(s), ans)

const romanToInt = function (s) {
  let total = 0
  let roms = s.split("")
  for (let i = 0; i < roms.length; i++) {
    const value = romNums[roms[i]]
    const nextValue = romNums[roms[i + 1]]
    console.log({ value, nextValue })
    if (value < nextValue) {
      total = nextValue - value
      console.log("1", total)
      i += 1
    } else {
      total += value
      console.log("2", total)
    }
  }
  return total
}

// TESTING

// s = "III"
// ans = 3
// console.log(romanToInt(s), ans)
// s = "IV"
// ans = 4
// console.log(romanToInt(s), ans)
// s = "IX"
// ans = 9
// console.log(romanToInt(s), ans)
// s = "LVIII"
// ans = 58
// console.log(romanToInt(s), ans)
s = "MCMXCIV"
ans = 1994
console.log(romanToInt(s), ans)
